crates/goose-cli/src/recipes/recipe.rs

```diff
@@ -0,0 +1,468 @@
+    if serde_json::from_str::<JsonValue>(content).is_ok() {
+        Ok(serde_json::from_str(content)?)
+    } else if serde_yaml::from_str::<YamlValue>(content).is_ok() {
+        Ok(serde_yaml::from_str(content)?)
+    } else {
+        Err(anyhow::anyhow!(
+            "Unsupported file format for recipe file. Expected .yaml or .json"
+        ))
+    }
+}
+
+fn extract_template_variables(template_str: &str) -> Result<HashSet<String>> {
+    let mut env = Environment::new();
+    env.set_undefined_behavior(UndefinedBehavior::Strict);
+
+    let template = env
+        .template_from_str(template_str)
+        .map_err(|_| anyhow::anyhow!("Invalid template syntax"))?;
+
+    Ok(template.undeclared_variables(true))
+}
+
+fn apply_values_to_parameters(
+    user_params: &[(String, String)],
+    recipe_parameters: Vec<RecipeParameter>,
+) -> Result<HashMap<String, String>> {
+    let mut param_map: HashMap<String, String> = user_params.iter().cloned().collect();
+    let mut missing_params: Vec<String> = Vec::new();
+    for param in recipe_parameters {
+        if !param_map.contains_key(&param.key) {
+            match (&param.default, &param.requirement) {
+                (Some(default), _) => param_map.insert(param.key.clone(), default.clone()),
+                (None, RecipeParameterRequirement::UserPrompt) => {
+                    let input_value = cliclack::input(format!(
+                        "Please enter {} ({})",
+                        param.key, param.description
+                    ))
+                    .interact()?;
+                    param_map.insert(param.key.clone(), input_value)
+                }
+                _ => {
+                    missing_params.push(param.key.clone());
+                    None
+                }
+            };
+        }
+    }
+    match missing_params.is_empty() {
+        true => Ok(param_map),
+        false => {
+            let formatted = missing_params
+                .iter()
+                .map(|key| format!("--params {}=your_value", key))
+                .collect::<Vec<_>>()
+                .join(" ");
+
+            Err(anyhow::anyhow!(
+                "Please provide the following parameters in the command line: {}",
+                formatted
+            ))
+        }
+    }
+}
+
+fn render_content_with_params(content: &str, params: &HashMap<String, String>) -> Result<String> {
+    // Create a minijinja environment and context
+    let mut env = minijinja::Environment::new();
+    env.set_undefined_behavior(UndefinedBehavior::Strict);
+    let template: Template<'_, '_> = env.template_from_str(content)
+        .map_err(|_| anyhow::anyhow!("Failed to render recipe, please check if the recipe has proper syntax for variables: eg: {{ variable_name }}"))?;
+
+    // Render the template with the parameters
+    template.render(params).map_err(|_| {
+        anyhow::anyhow!(
+            "Failed to render the recipe - please check if all required parameters are provided"
```

Comment:
can you capture the error in `map_err(|e|...` and add the error from jinja to the anyhow message